ECS

Akashae98 opened the pull request:
Classes:
- Entity: hashmap of components, methods for add/remove/search 
- Component (abstract)
- PositionComponent, VelocityComponent, SpriteComponent, ColliderComponent
- System (abstract): update method
- RenderSystem, MovementSystem

Commit:
ECS structure: add classes 
- Entity: hashmap of components, methods for add/remove/search 
- Component (abstract)
- PositionComponent, VelocityComponent, SpriteComponent, ColliderComponent
- System (abstract): update method
- RenderSystem, MovementSystem

Commit:
add method has component on Entity class

Commit:
add logic in method update of MovementSystem and RenderSystem. 
add logic in SpriteComponent

Commit:
feat: render system implements flip and provisional rotation
- Updated MovementSystem, RenderSystem, SpriteComponent, ColliderComponent
- Adds flip property to SpriteComponent
- MovementSystem calculates orientation based on velocity and determines flip on collisions
- RenderSystem handles flip/rotation
- ColliderComponent adds bounce detection

Commit:
test: add coral fishes as entities of ECS

gcervifi commented on src/main/java/com/mycompany/animacionpecera/MovementSystem.java:
```diff
@@ -0,0 +1,92 @@
+
+            }
+            //To collide or rotate..
+            if (entity.hasComponent(ColliderComponent.class) && entity.hasComponent(SpriteComponent.class)) {
```
Mueve esto a un sistema separado. Movement se encarga de cambiar posiciones. Haz otro sistema que se encargue de interactuar con el borde de la pecera y cada elemento.

gcervifi commented on src/main/java/com/mycompany/animacionpecera/SpriteComponent.java:
```diff
@@ -0,0 +1,36 @@
+    public double rotation;
+    public boolean flip = false;
+
+    public SpriteComponent(String imagePath, double size) {
```
Crea una clase que se encargue de cargar en memoria las imagenes necesarias. Sprite componente solo ha de tener el path a la imagen o una string que identifique dicha imagen idealmente. Dicha clase se encargara de cargar las imagenes y dejarlas en memoria para que RenderSystem pueda obtenerlas.

gcervifi commented on src/main/java/com/mycompany/animacionpecera/GameLoop.java:
```diff
@@ -58,7 +64,7 @@ public void handle(long now) {
+            //System.out.println("Skipping frame: " + deltaTime);
             lastUpdate = now;
             return;
         }
```
Mueve la logica de  LinearGradient background dentro de renderSystem. La gracia del render system, es que podrias ejecutar tu pecera sin necesidad de renderizar nada, pero calculando la logica de los peces.

gcervifi commented:
Crea una carpeta llamada Components y otra Systems. Pon las clases correspondientes.

Akashae98 commented on src/main/java/com/mycompany/animacionpecera/RenderSystem.java:
```diff
@@ -0,0 +1,53 @@
+                gc.drawImage(sprite.image, -sprite.getWidth() / 2,
+                        -sprite.getHeight() / 2, sprite.getWidth(), sprite.getHeight());
+
+                gc.restore();
```
sii, lo he estado usando en la rama fishidle para aplicar tonos distintos a los peces

Commit:
feat: add Transfom class with position, scale and rotation using Affine

Commit:
feat: add ImageManager class with HashMap(images remained in memory)

Commit:
refactor:
-System package for systems.
-Extract rotation and rebound from MovementSystem to ColliderSystem using Transform.
-Rename class System to GameSystem.

Commit:
refactor: 
-Package of Components.
-SpriteComponent uses ImageManager.
-Remove use of boundingbox.

Commit:
feat: Component Bubble

Commit:
method update uses list of Systems for ECS

Commit:
Refactor for entities using Transform component

Commit:
Format changes

Commit:
add solar rays to the background

gcervifi commented on src/main/java/com/mycompany/animacionpecera/Components/Transform.java:
```diff
@@ -0,0 +1,86 @@
+ */
+public class Transform extends Component {
+
+    private double x, y, rotation, scaleX, scaleY;
```
Ponlo en varias lineas

gcervifi commented on src/main/java/com/mycompany/animacionpecera/Components/Transform.java:
```diff
@@ -0,0 +1,86 @@
+    }
+
+    public javafx.scene.transform.Affine getAffineTransform() {
+        javafx.scene.transform.Affine affine = new javafx.scene.transform.Affine();
```
No uses affine en el componente

gcervifi commented on src/main/java/com/mycompany/animacionpecera/Components/SpriteComponent.java:
```diff
@@ -0,0 +1,32 @@
+    public boolean flip = false;
+
+    public SpriteComponent(String imagePath, double size) {
+        this.image = ImageManager.getInstance().getImage(imagePath);
```
NO guardes la imagen aqui. Aqui solo tiene que tener el path de la imagen. El imageManager se usa en el RenderSystem.

Akashae98 replied:
okey, y puedo dejar los métodos de getwidth y getHeight si uso el image manager?

Akashae98 replied:
sii era por eso, el size en transform por que? por si luego cambia el tamaño?

Akashae98 replied:
no tiene sentido que el sprite tenga size entonces no?

gcervifi replied:
Transform ha de tener scale. Scale afecta a todos los componentes. EL size de collition y sprite definen sus atributos, que luego se veran afectados por scale.

gcervifi commented on src/main/java/com/mycompany/animacionpecera/System/RenderSystem.java:
```diff
@@ -0,0 +1,134 @@
+    @Override
+    public void update(List<Entity> entities, double deltaTime) {
+        // Paint the background
+        gc.setFill(background);
```
Añade un metodo para borrar todo el contenido del canvas.

gcervifi commented on src/main/java/com/mycompany/animacionpecera/System/RenderSystem.java:
```diff
@@ -0,0 +1,134 @@
+            };
+            
+            // Vertical gradient for the ray
+            LinearGradient rayGradient = new LinearGradient(
```
Mueve el gradiente y los rayos a una funcion aparte

gcervifi commented on src/main/java/com/mycompany/animacionpecera/System/ColliderSystem.java:
```diff
@@ -0,0 +1,89 @@
+ *
+ * @author carol
+ */
+public class ColliderSystem extends GameSystem {
```
Cambia el nombre de este sistema a otro mas especifico.

gcervifi commented on src/main/java/com/mycompany/animacionpecera/GameLoop.java:
```diff
@@ -64,10 +67,10 @@ public void handle(long now) {
 
         // Rendering
-        renderScene(deltaTime);
+        //renderScene(deltaTime);
```
Borra las funciones comentadas.
Crea un nuevo system DebugCollisionRender. Y gestiona ahi el pintar el collider.

gcervifi commented on src/main/java/com/mycompany/animacionpecera/GameLoop.java:
```diff
@@ -77,15 +80,18 @@ public void handle(long now) {
-    private void updateGameLogic(double deltaTime) {
-        for (SceneObject object : sceneObjectList) {
-            object.move(deltaTime);
+    private void update(double deltaTime) {
```
Ponlo directamente en el game loop, esta funcion a parte no aporta nada.

Commit:
changes PR

Commit:
changes PR: added method in RenderSystem to clean the canvas.
Attribute size in Transform.
Sprite stores the imageKey for the ImageManager.
Methods getWidth() and getHeight() now on the ImageManager

gcervifi commented on src/main/java/com/mycompany/animacionpecera/Components/Transform.java:
```diff
@@ -0,0 +1,79 @@
+
+    private double x, y; 
+    private double rotation;
+    private double scaleX, scaleY;
```
Scale YA es size, estas duplicando la informacion.

Akashae98 replied:
creía que eso era para escalar el tamaño

Commit:
changes PR: using scale instead of size

Commit:
feat: add new system DebugCollisionRender for painting the colliders

Akashae98 committed:
test: colliders and image using different scale x and y.
refactor SpriteComponent without scale.

Akashae98 commented:
Pero tengo que aplicarlo en los sistemas, no en el componente no?

El vie, 5 sept 2025, 13:39, Carol Vilar ***@***.***> escribió:

> Ah vale se lo estoy aplicando en el render del collider pero luego para el
> collider lógico no
>
> El vie, 5 sept 2025, 12:49, gcervifi ***@***.***> escribió:
>
>> *gcervifi* left a comment (Akashae98/pecera#35)
>> <https://github.com/Akashae98/pecera/pull/35#issuecomment-3257924215>
>>
>> Para el render, si usa afine, pero el collider debe tambien usar el
>> scale. Y tienes que multiplicarlo, sino no cuadrara el sprite con el
>> collider.
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/Akashae98/pecera/pull/35#issuecomment-3257924215>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/BKS42ROY7C5OTKGCJFZ6X2L3RFTCTAVCNFSM6AAAAACEPF5AJ6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTENJXHEZDIMRRGU>
>> .
>> You are receiving this because you authored the thread.Message ID:
>> ***@***.***>
>>
>

Commit:
ajusted scale of the collider in CanvasBounceAndRotationSystem

Akashae98 commented:
Pues no sé si estará bien así porque he terminado modificando el collider igualmente xD, ya me dices

Commit:
feat: convert ColliderComponent to abstract class, add CircleCollider and 
BoxCollider, refactor systems of colliders

gcervifi commented on src/main/java/com/mycompany/animacionpecera/System/CanvasBounceAndRotationSystem.java:
```diff
@@ -0,0 +1,93 @@
+        double bottomEdge = collider.getBottom(transform);
+
+        // Verify collisions on left and right
+        if ((leftEdge <= 0 && velocity.velX < 0) || (rightEdge >= canvasWidth && velocity.velX > 0)) {
```
Mucho cuidado, porque no es los mismo comparar un rectangulo sin rotacion que uno con rotacion. Un rectangulo no rotado seria un AABB y su calculo es facil. Si esta rotado, tienes que hacer mas calculos.

Akashae98 commented:
Ah vale, no se lo quería poner pero pensaba que tenía que heredarlo al
extender del padre xD

El sáb, 6 sept 2025, 11:49, gcervifi ***@***.***> escribió:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In
> src/main/java/com/mycompany/animacionpecera/Components/CircleCollider.java
> <https://github.com/Akashae98/pecera/pull/35#discussion_r2326730336>:
>
> > @@ -0,0 +1,41 @@
> +/*
> + * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
> + * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
> + */
> +package com.mycompany.animacionpecera.Components;
> +
> +/**
> + *
> + * @author User
> + */
> +public class CircleCollider extends ColliderComponent {
> +
> +    public CircleCollider(double width, double height) {
>
> Hija mia, cuando has visto que un circulo tenga altura?
> Quita width, Height de Collider padre.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Akashae98/pecera/pull/35#pullrequestreview-3191974801>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BKS42RJ4KRWEU2QJZNHF4T33RKUYZAVCNFSM6AAAAACEPF5AJ6VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZTCOJRHE3TIOBQGE>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

Akashae98 committed:
refactor constructor of CircleCollider uses double diameter

Akashae98 committed:
refactor handleBounce doesn't return a boolean of rebound

Commit:
feat: new VelocitySystem manage rotation, refacctor CanvasBounceSystem

Akashae98 committed:
feat: new component for multicolor fishes

Akashae98 committed:
feat: new component for multicolor fishes

Akashae98 committed:
refactor: component MulticolorFish manages ColorAdjust, extract it from RenderSystem

Akashae98 committed:
format changes

Akashae98 committed:
add red_fish image to method addMulticolorFish
 + format changes

gcervifi commented on src/main/java/com/mycompany/animacionpecera/Components/MulticolorFish.java:
```diff
@@ -0,0 +1,68 @@
+    Random random = MainScene.random;
+
+    public MulticolorFish(Color color) {
+        this.randomHue = random.nextDouble();
```
No pongas random en el componente. Pon todo lo que necesites en el contructor. Es el encargado de crear los Componentes el que decidira aleatoriamente los valores, no el componente en si.

gcervifi replied:
Es mas, como deberia ir esto es:
EL componente debería llamarse ColorSprite.
Los atributos del componente son Hue, Saturation, Y Brightness.
Crea una clase llamada ColorUtils, pon toda la logica de crear colores ahi. 

Los componentes solo tienen datos no logica.
Con el nombre de ColorSprite, ese componente es agnostico, no es para peces, es para sprites. 
Siempre que puedas, los componentes tienen que servir para todas las entidades, no para una en concreto.

gcervifi commented on src/main/java/com/mycompany/animacionpecera/System/VelocitySystem.java:
```diff
@@ -0,0 +1,45 @@
+ */
+public class VelocitySystem extends GameSystem {
+
+    public VelocitySystem() {
```
No necesitas definit contructor si esta vacio

gcervifi commented on src/main/java/com/mycompany/animacionpecera/System/RenderSystem.java:
```diff
@@ -60,7 +58,7 @@ public void update(List<Entity> entities, double deltaTime) {
 
-            //a white circle to highlight the bubbles..?
+            //a white circle to highlight the bubbles…?
             if (entity.hasComponent(Bubble.class)) {
```
Cambia el render de burbuja a un sprite, asi no tienes que hacer un caso especial para burbuja, funcionaria igual para peces que burbujas.

Akashae98 commented:
Guay!! Una cosa, habrá que borrar las clases del anterior sistema no?
Estaba pensando que para que quede visible la arquitectura que usaba
sceneobject y clases abstractas se podría dejar en una carpeta como
versión. Osea ponerla en la raíz de la carpeta de la repo, no?? y le podría
poner un readme a parte explicando la jerarquía, cómo funciona,  que usa la
inyección de dependencias o de clases abstractas, etc. Así queda reflejado
lo que he ido haciendo, porque no creo que nadie me vaya a ir a mirar las
ramas.

El sáb, 13 sept 2025, 13:38, gcervifi ***@***.***> escribió:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/main/java/com/mycompany/animacionpecera/System/RenderSystem.java
> <https://github.com/Akashae98/JavaFX-Fish-Simulation/pull/35#discussion_r2346443485>
> :
>
> > @@ -60,7 +58,7 @@ public void update(List<Entity> entities, double deltaTime) {
>              //to manually draw the oval, we need to apply the scale without affine
>              double widthDraw = image.getWidth() * transform.getScaleX();
>
> -            //a white circle to highlight the bubbles..?
> +            //a white circle to highlight the bubbles…?
>              if (entity.hasComponent(Bubble.class)) {
>
> Cambia el render de burbuja a un sprite, asi no tienes que hacer un caso
> especial para burbuja, funcionaria igual para peces que burbujas.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Akashae98/JavaFX-Fish-Simulation/pull/35#pullrequestreview-3220385647>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BKS42RNCGMAGGEEY2TMYS7L3SP65FAVCNFSM6AAAAACEPF5AJ6VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZTEMRQGM4DKNRUG4>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

Akashae98 committed:
refactor: rename MulticolorFish component to ColorSprite which manage the ColorEffect.
Also extract the logic of the component into ColorUtils.

Akashae98 committed:
delete classes of SceneObjects

Akashae98 committed:
delete empty constructor

Akashae98 committed:
add to gitignore the .run.xml of Intellij

Akashae98 committed:
Upgrade JavaFX and Java Target:

- Updated JavaFX from 21 to 25
- Updated maven.compiler.source from 21 to 25
- Updated maven.compiler.target from 21 to 25
- Updated maven-compiler-plugin release from 21 to 25

All configuration updated for Java 25 LTS compatibility.

Akashae98 committed:
Update JavaFX project to Maven wrapper and refresh docs

Akashae98 committed:
Move run.ps1 script for Maven to the directory root

Akashae98 committed:
Merge branch 'upgrade-java-version' into ECS_#33

carolvilar-art committed:
Add automated desktop release workflow

Akashae98 committed:
Fix GitHub release publish step

Akashae98 committed:
Fix desktop packaging and add Windows fallback launcher

Akashae98 committed:
Add Linux fallback launcher
